#51 - On-Boarding UX

noahspoke opened the pull request:
This is the new PR, so that I can own it and also request a review from you :)

noahspoke commented:
@datracka Please review this as soon as you are able. Once we know this is working, we can merge into staging and get this tested in a live environment.

datracka commented on app/client/epics/searches.epics.js:
```diff
@@ -25,7 +26,8 @@ const saveSearchEpic = (action$, state$, { fetch$ }) =>
         headers: {
-          contentType: "application/json"
+          contentType: "application/json",
+          Authorization: "Bearer " + getValidToken()
```
good but now, actually, ALL requests (with the exception of Sign In Sign Up) should have auth token so I would go for a more scalable solution. I am thinking to create wrapper and populate `headers` with default behaviour. and pass it as dependency. Ideas are welcome. I can help with it if you like

datracka replied:
Let's leave as it is for now.  Im next branch we do a proper endpoints securitization.

datracka commented on app/client/epics/searches.epics.js:
```diff
@@ -68,7 +70,8 @@ const getDonationsEpic = (action$, state$, { fetch$ }) =>
           Accept: "application/json",
-          "Content-Type": "application/json"
+          "Content-Type": "application/json",
+          Authorization: "Bearer " + getValidToken()
```
same as above

datracka commented on app/client/epics/searches.epics.js:
```diff
@@ -99,7 +102,8 @@ const getCitiesEpic = (action$, state$, { fetch$ }) =>
           Accept: "application/json",
-          "Content-Type": "application/json"
+          "Content-Type": "application/json",
+          Authorization: "Bearer " + getValidToken()
```
same as above

datracka commented on app/client/epics/user.epic.js:
```diff
@@ -75,4 +77,38 @@ const signInEpic = (action$, state$, { fetch$ }) =>
+        responseType: "json"
+      }).pipe(
+        map(response => {
+          if (resolve) resolve(response.response);
```
all this boilerplate is for handling server form validation (email / pass). Sign out does not need all this. I can help with it if you have questions

datracka replied:
I changes in the https://github.com/BravoGroup/PADonors/pull/103 I was telling you

datracka commented on app/client/epics/user.epic.js:
```diff
@@ -75,4 +77,38 @@ const signInEpic = (action$, state$, { fetch$ }) =>
+    })
+  );
+
+export { signUpEpic, signInEpic, signOutEpic };
```
btw, do we have server form validation? I did not see any implementation there. You can get inspired in Signup

datracka commented on app/client/modules/App.js:
```diff
@@ -24,7 +24,7 @@ const App = props => {
         <ContextProvider value={{ ...props }}>
           <Provider store={configureStore()}>
-            <Routes />
+            <Routes currentUser={props.currentUser} />
```
I would not do that. First routes has nothing to do with states. Secondly you have this information in Context so you can grab it from any place. check it out https://reactjs.org/docs/context.html and get inspired by `Filters.js` (how we consume context)

datracka commented on app/client/routes.js:
```diff
@@ -21,36 +23,58 @@ const history = createBrowserHistory();
 );
 
-export default compose(connect(prop("userReducer")))(Routes);
+export default compose(
```
not here, it should be in `App` IMO.

datracka replied:
moved to configureStore as described here https://github.com/BravoGroup/PADonors/pull/103

datracka commented on app/client/routes.js:
```diff
@@ -21,36 +23,58 @@ const history = createBrowserHistory();
-      <Route path="/employer/:id" component={Employer} />
-      <PrivateRoute path="/saved-searches" component={SavedSearches} />
-      <PrivateRoute path="/profile" component={Profile} />
+      <PrivateRoute exact path="/" component={Welcome} loggedIn={loggedIn} />
```
Ok, you added all private routes in this PR.... We should change default path to signIn (instead Welcome). Let's talk about it.

noahspoke replied:
For now though, let's leave this as is due to the testing nature of the product. That means any user that is not logged in will just be redirected to `/signin`.

datracka commented on app/client/routes.js:
```diff
@@ -21,36 +23,58 @@ const history = createBrowserHistory();
-const Routes = () => (
+const setCurrentUser = ({ currentUser, signInSuccess }) => {
+  if (currentUser) {
+    signInSuccess({
```
I don't think so use an action related to signIn for inilialise the state (that is what you want to do)
check it out this: https://redux.js.org/recipes/structuringreducers/initializingstate

Nevertheless we have to talk about all this.

noahspoke replied:
Thanks for posting the link, that helped clarify!

datracka commented on app/client/state/user.duck.js:
```diff
@@ -21,15 +19,28 @@ export const { actions, initial, selectors, reducer } = autoduck({
+    signInSuccess: (state, { user, token }) => {
+      return {
+        ...state,
+        user,
```
remember we were discussing about remove user from state... ok, we can leave it as we will use in a later state for profile. 
Nevertheless I noticed that when is updated from parameter templates the user value is `stringified` and invalid

datracka commented on app/client/state/user.duck.js:
```diff
@@ -21,15 +19,28 @@ export const { actions, initial, selectors, reducer } = autoduck({
-    signOut: state => state,
-    signOutSuccess: state => state,
+    signOut: state => {
+      console.log("SIGN OUT IN DUCK");
```
remove console log

datracka commented on app/client/state/user.duck.js:
```diff
@@ -21,15 +19,28 @@ export const { actions, initial, selectors, reducer } = autoduck({
+      return state;
+    },
+    signOutSuccess: state => {
+      console.log("SIGN OUT SUCCESS: " + JSON.stringify(state));
```
remove console.log

datracka commented:
I added a label 'ready for review' I am a little nerd with this kind of stuff whahahaha! 👯

datracka commented:
@Noah Regarding issue described above “if App Token from local storage is removed manually when user still is signed In application is not longer responding.”:

The problem is that sign_out end points rejects request if JWT is not present. Then devise cookie session never is deleted (besides the token is not revoked but it does not matter because the client does not have it anymore)

Sign out should work for delete devise session even without token. 

Nevertheless we can talk and go for ONLY a token solution. And rely in local storage for checking if user is loggedIn or not. redux state would have only user information if loaded by API endpoint...
I gonna make a diagram to make clearer PROS and CONS ok?

datracka commented:
Hi @noah as promised here 2 diagrams to clarify user management. 

First one is the minimum implementation possible that we need right now. is 100% based in JWT token and NO REDUX state at all. Only we relay in Local Storage if token exists or not. 

<img width="1004" alt="screenshot 2018-11-07 at 09 42 35" src="https://user-images.githubusercontent.com/3977088/48120226-d8b93b80-e271-11e8-9301-f8550b2fce78.png">

**>> In red missing features to implement**

If we agree this implementation. You should rollback / change some of the current implementation (but not so much I think) the most important part is disable cooke server session for now.  https://stackoverflow.com/questions/31641453/devise-custom-strategies-still-setting-session-cookies
**In this implementation we don't have any user data in client. But we don't need that in this stage so Why implement it right now?**

If you still want to go for a "full house" implementation like this: 

<img width="994" alt="screenshot 2018-11-07 at 09 42 45" src="https://user-images.githubusercontent.com/3977088/48120455-76ad0600-e272-11e8-939f-0d936de5d4e9.png">

Then we have to be very careful about what is our source or truth for checking when user is logged or not. It is not easy to handle at all! So it will make it live more difficult. (and again we don't need this in this stage of the application)

Last, sorry all headaches I brought to you with the decision of using JWT token authentication instead classical cookie session server. I think it the way to go but of course I am open to discuss whatever makes you more comfortable.

noahspoke commented:
> Hi @noah as promised here 2 diagrams to clarify user management.
> 
> First one is the minimum implementation possible that we need right now. is 100% based in JWT token and NO REDUX state at all. Only we relay in Local Storage if token exists or not.
> 
> <img alt="screenshot 2018-11-07 at 09 42 35" width="1004" src="https://user-images.githubusercontent.com/3977088/48120226-d8b93b80-e271-11e8-9301-f8550b2fce78.png">
> 
> **>> In red missing features to implement**
> 
> If we agree this implementation. You should rollback / change some of the current implementation (but not so much I think) the most important part is disable cooke server session for now. https://stackoverflow.com/questions/31641453/devise-custom-strategies-still-setting-session-cookies
> **In this implementation we don't have any user data in client. But we don't need that in this stage so Why implement it right now?**
> 
> If you still want to go for a "full house" implementation like this:
> 
> <img alt="screenshot 2018-11-07 at 09 42 45" width="994" src="https://user-images.githubusercontent.com/3977088/48120455-76ad0600-e272-11e8-939f-0d936de5d4e9.png">
> 
> Then we have to be very careful about what is our source or truth for checking when user is logged or not. It is not easy to handle at all! So it will make it live more difficult. (and again we don't need this in this stage of the application)
> 
> Last, sorry all headaches I brought to you with the decision of using JWT token authentication instead classical cookie session server. I think it the way to go but of course I am open to discuss whatever makes you more comfortable.

For now, I will focus on the basic implementation of the JWT token in `localStorage` and eventually we will migrate to a full implementation.

noahspoke commented:
This new implementation solely focuses on the JWT tokens, which I think is the direction this is all headed haha. Correct me if I'm wrong :)?

datracka commented on app/client/components/common/AppBar.js:
```diff
@@ -200,6 +203,20 @@ export default compose(
       this.props.setState({ anchorProfile: null });
+    },
+    handleSignOut: props => ({ signOut }) => {
+      new Promise((resolve, reject) => {
```
I applied the changes in this https://github.com/BravoGroup/PADonors/pull/103
I think you will get what I meant...

noahspoke commented:
@datracka - search is now working on my end. Try running `DROP TABLE user_searches` and re-run the migrations. That is what worked for me, so that the column references are up to date in `search_controllers.rb`

datracka commented:
thanks @noah! 

I merged from staging for the last test to see if all works properly! It looks almost perfect only a couple of things:

- When Sign in and user does not exist no message pops up... it is bad UX. I will fix for myself it this afternoon ok? for me should be easier as I just implemented something similar for sign up.

- `user_searches` table inserts a row for each new search regardless if it is exactly the same search. I think `user_id` and `search_id` should be unique too to avoid.. .I think it has no value to store how many times (rows) a user performs a search and in any case we could set a counter field to avoid to bloat database... what are your thoughts there?

noahspoke commented:
@datracka I think we are finally ready to merge this! 😆The new validations are working great, and the layout looks awesome. Since I am the owner, I cannot approve this PR but I am going to merge this anyways since you've seen the changes. (All I've added is a merge from `staging` and updated `manifest.json`. I'm going to post this same thing in Slack. Wooo!
